index/scorch/persister.go

```diff
@@ -245,6 +245,14 @@ func notifyMergeWatchers(lastPersistedEpoch uint64,
 	return watchersNext
 }
 
+func (s *Scorch) getNumSnapshotsToKeep(ourSnapshot *IndexSnapshot) int {
+	if ourSnapshot != nil &&
+		ourSnapshot.creator != forceMergeCreator {
+		return s.numSnapshotsToKeep
+	}
+	return 1
```

Comment:
Wondering if violating the numSnapshotsToKeep contract here in case of forced merge is OK?
I may be missing something obvious here, but this is something you're not doing below within pausePersisterForMergerCatchUp ?

Reply:
Indeed, the reason we are doing it here is that - Usually a full compaction will need 2X++ size requirement. In other words, we don't want to keep multiple copies of the same data.. So this is deliberate decision.

Reply:
I share the concern, and don't understand the rationale.  What about a force merge implies that I want to violate the regular settings?

Reply:
Let me try to explain this with a merge policy for a single index file,
In this case, the entire data is getting rewritten to a single new file. So if we are keeping the older N snapshots, then this would result in a lot of space consumption due to data duplication. Also during merge phase - the disk consumption was going very high.

This extra space cost is  hard to justify when a customer try force merge for some potential benefits out it.

Reply:
But you're using phrase like "a lot of space", but what does that mean?  Maybe the user understands their space requirements and planned accordingly, but you've already decided for them to ignore the configuration that they specified.

It seems like your real complaint is that N snapshots != N useful snapshots.  Forced merge/compaction often results in useless snapshots.  I would prefer we try to address that issue separately, or at least make this change in behavior optional.

Reply:
I used the term a "lot of" because during my experiments, I saw the disk space spiking upto 3X or even more during the merge process. But eventually it dips to 2X or some bigger value. The eventual file size partly could be attributed to larger segment bloating issue. And as of today, a single segment merge was spiking the space usage more than 2X. 

The wrinkle I see(non convincing part) here is that - people often associate compaction/forceMerge with an index size reduction/space savings and such bigger requirements of space is contradicting that.

Secondly, another thought popping to mind while writing this is that (may be I have lost the rollback fundamentals from my mind)  keeping 3/N number of snapshots isn't a a big win after all. For eg: in cbft at **max** each batch/snapshot can contain only 200 mutations. 

So with the default snapshot keep count of 3 will save 600(3X200) mutations.

Faster rollback ability to 600 mutation isn't a huge win considering the time to re-index it?
So don't think its a significant trade off to drop older snapshots here..

Making this configurable is certainly an option too..

Reply:
So, in normal usage, keeping 3 snapshots should not mean anything close to 3x the size of the index.  Snapshots share older segments, and so in many typical cases, I would expect much of the data of the index to be shared in older segments that haven't changed enough to be rewritten.  Only a handful of segments would differ, and thus be kept around.  These would tend to be newer/smaller segments anyway.  There are of course other use cases where you're constantly modifying the same documents over and over, and yes in that case 3 old snapshots might mean having 3 complete sets of segments representing the data, and thus 3x the data.  However, I would not expect that to be common.

It's also possible that our current known issue of older/larger segments being less space efficient actually undermines this math, and is contributing to confusion.

Next, I would agree with you that generally people would expect *the result* of a merge/compaction to possibly save space.  I also think they would expect that *during* the operation it could in fact consume 2x more.  I think with appropriate documentation and explanation about the way merging works would convince most people.

I'm still holding on to my main objection.  The user requested N old snapshots in their configuration.  And this code intentionally ignores that configuration, for the stated reason of not taking too much space (even though the code has no idea how much space that is, or if the user has planned appropriately).  As a user, when code silently does something different than I asked, I find that to be a problem.

Reply:
I agree that overriding the `num snapshots to keep` is a side effect of this API, but my plan was to document that in the API specs. As the rewards there(by keeping 3/N older snapshots) anyway is of lesser significance to the user.

But any extra space mandates from the API call has a higher cost factor to take care of which might needs cluster size planning Or given the un bounded nature of size requirements as of today- we can't rule out the possibility of pushing them to buy more hardware, perhaps?. This concern outweighs the former for me.

I also agree that the effectiveness(return on investment) of existing `num snapshots to keep` feature is completely separate topic. 
But since we already touched that here-  I wasn't saying the upper bound was going to be strictly 3X or I didn't try to quantify that earlier during my trials with this API.

Even in a normal scenario,
If a single mutation gets indexed,  it could create a newer disk segment file, which could potentially trigger a merge operation. Merge can invalidate any random N number of disk files from different size tiers to result in a huge size duplication too.. (or there is no assurance that merge will happen only at the lower/bottom tier of smaller segments?)  
And let's say - if a single merge operation involved 1 older segment of gigantic size, then we will end up creating lot of space waste out of a single mutation ingestion. 

So on an average, the extra space occupied by these reasonably sized shared disk segments aren’t justifying the planned meagre performance boost for the seldom(debatable) occurring rollback that to upto a maximum point in history of a few mutations.?